prover/src/opcodes/groestl.rs

```diff
@@ -0,0 +1,815 @@
+        let src2_vals = from_fn(|i| table.add_committed(format!("src2_val_{i}")));
+
+        // Get the base address for the first and second source values.
+        let src1_abs =
+            table.add_computed("src1_addr_0", state_cols.fp + upcast_col(state_cols.arg1));
+        let src2_abs =
+            table.add_computed("src2_addr_0", state_cols.fp + upcast_col(state_cols.arg2));
+
+        // Get all the addresses for the first and second source values.
+        let mut src1_addresses = [src1_abs; 16];
+        for (i, s1) in src1_addresses.iter_mut().enumerate().skip(1) {
+            *s1 = table.add_computed(format!("src1_addr_{i}"), src1_abs + B32::from(i as u32));
+        }
+        let mut src2_addresses = [src2_abs; 16];
+        for (i, s2) in src2_addresses.iter_mut().enumerate().skip(1) {
+            *s2 = table.add_computed(format!("src2_addr_{i}"), src2_abs + B32::from(i as u32));
+        }
+
+        // Transpose src1 and src2 values to get the correct B32 lookups in the VROM.
+        let src1_transposition = TransposeColumns::new(&mut table, src1_vals);
+        let src2_transposition = TransposeColumns::new(&mut table, src2_vals);
+        let src1_vals_packed = src1_transposition.output;
+        let src2_vals_packed = src2_transposition.output;
+
+        // Pull the first and second source values from the VROM channel.
+        for i in 0..16 {
+            table.pull(vrom_channel, [src1_addresses[i], src1_vals_packed[i]]);
+            table.pull(vrom_channel, [src2_addresses[i], src2_vals_packed[i]]);
+        }
+
+        // p_state_in = src1_val XOR src2_val.
+        let p_state_in: [Col<binius_field::BinaryField8b, 8>; 8] =
+            from_fn(|i| table.add_computed("state_in", src1_vals[i] + src2_vals[i]));
+
+        let p_op = Permutation::new(
+            &mut table,
+            binius_m3::gadgets::hash::groestl::PermutationVariant::P,
+            p_state_in,
+        );
+
+        // p_out = P(p_state_in)
+        let p_out_array = p_op.state_out();
+
+        // Carry out the Q permutation.
+        let q_op = Permutation::new(
+            &mut table,
+            binius_m3::gadgets::hash::groestl::PermutationVariant::Q,
+            src2_vals,
+        );
+
+        // q_out = Q(src2_val)
+        let q_out_array = q_op.state_out();
+
+        // interm = p_out XOR src1_val.
+        let interm: [Col<B8, 8>; 8] =
+            from_fn(|i| table.add_computed(format!("interm_{i}"), p_out_array[i] + src1_vals[i]));
+
+        // Compute the final output: out = Q(m) XOR P(src1_val XOR src2_val) XOR
+        // out = interm XOR q_out.
+        let out: [Col<B8, 8>; 8] =
+            from_fn(|i| table.add_computed(format!("out_{i}"), interm[i] + q_out_array[i]));
+
+        // TODO: Adapt the gadget to make the output shape more generic.
+        // We transpose the output. We do not use the transposition gadget as the output
+        // shape is a bit different.
```

Comment:
I think you should also describe from high-level what transformation are you applying